app/src/main/java/com/weather/weathercheckerapp/ViewModelFactory.kt

```diff
@@ -0,0 +1,25 @@
+package com.weather.weathercheckerapp
+
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.ViewModelProvider
+import javax.inject.Inject
+import javax.inject.Provider
+import javax.inject.Singleton
+
+@Suppress("UNCHECKED_CAST")
+@Singleton
+class ViewModelFactory @Inject constructor(private val viewModelsMap: Map<Class<out ViewModel>, @JvmSuppressWildcards Provider<ViewModel>>) : ViewModelProvider.Factory {
+
+    override fun <T : ViewModel> create(modelClass: Class<T>): T {
```

Comment:
I think this function can be simplified. You don't need to catch the exceptions.
You don't need to store the creator, you can return it immediately.
Also the elvis operator (`?:`) usually goes the the beginning of a line, not the end of it.
You can utilize the safe cast operator as well: `as?`
Something like this:
´´´
return (viewModelsMap[modelClass] 
        ?: viewModelsMap.asIterable()
                .firstOrNull { modelClass.isAssignableFrom(it.key)  }
                ?.value)
        ?: get() as? T
        ?: throw IllegalArgumentException("unknown model class $modelClass")
´´